### PR TITLE
Remove Constant Lock and Unlocking in Ads

### PR DESCRIPTION
**Pull request by adiprerepa**
I am clueless as to why we are constantly Locking and Unlocking in `shouldRespond()` - it really annoyed me. This just locks and unlocks once.

### CODE REVIEW

**Comment by adiprerepa** [comment]
the only possible upside i can see to keeping the current config is so the code can selectively `Lock()` or `RLock()`, but it seems that the tradeoff between readability and nano-optimizations needs to be made here.

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -280,6 +280,9 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream discovery.AggregatedD
 func (s *DiscoveryServer) shouldRespond(con *Connection, request *discovery.DiscoveryRequest) bool {
 	stype := v3.GetShortType(request.TypeUrl)
 
+	con.proxy.RLock()
```
Comment (stevenctl) [inline pilot/pkg/xds/ads.go:283]: There are multiple writes here, the delete on 301, and a map write on 308

**Comment by ramaraochavali** [comment]
The reason for selective lock and unlock is to reduce the concurrency by taking lock at the beginning as this is a frequently used method. But I have n't measured the diff to say one where or other. If you are making this change, can you possibly write a simple benchmark and compare before/after?

**Pull request closed without merging**